narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,233 +3,404 @@
+) -> DagResult<()> {
+    // Send request to fetch certificates.
+    let request = FetchCertificatesRequest {
+        exclusive_lower_bounds: committed_rounds.into_iter().collect(),
+        max_items: MAX_CERTIFICATES_TO_FETCH,
+    };
+    let response =
+        fetch_certificates_helper(&state.name, &state.network, &committee, request).await;
+
+    // Process and store fetched certificates.
+    let num_certs_fetched = response.certificates.len();
+    process_certificates_helper(response, &state.tx_certificates_loopback).await?;
+    state
+        .metrics
+        .certificate_waiter_num_certificates_processed
+        .with_label_values(&[&committee.epoch().to_string()])
+        .add(num_certs_fetched as i64);
+
+    debug!("Successfully fetched and processed {num_certs_fetched} certificates");
+    Ok(())
+}
+
+/// Fetches certificates from other primaries concurrently, with ~5 sec interval between each request.
+/// Terminates after the 1st successful response is received.
+#[instrument(level = "debug", skip_all)]
+async fn fetch_certificates_helper(
+    name: &PublicKey,
+    network: &P2pNetwork,
+    committee: &Committee,
+    request: FetchCertificatesRequest,
+) -> FetchCertificatesResponse {
+    trace!("Start sending fetch certificates requests");
+    let request_interval = Duration::from_secs(5);
+    let mut peers: Vec<NetworkPublicKey> = committee
+        .others_primaries(name)
+        .into_iter()
+        .map(|(_, _, network_key)| network_key)
+        .collect();
+    loop {
+        peers.shuffle(&mut ThreadRng::default());
+        let mut fut = FuturesUnordered::new();
+        for peer in peers.iter() {
+            fut.push(network.fetch_certificates(peer, request.clone()));
+            let mut interval = Box::pin(time::sleep(request_interval));
+            tokio::select! {
+                res = fut.next() => match res {
+                    Some(Ok(resp)) => {
+                        return resp;
+                    }
+                    Some(Err(e)) => {
+                        debug!("Failed to fetch certificates: {e}");
+                        // Issue request to another primary immediately.
+                    }
+                    None => {}
+                },
+                _ = &mut interval => {
+                    debug!("fetch_certificates_helper: no response within timeout. Sending out a new fetch request.");
+                }
+            };
+        }
+    }
+}
+
+#[instrument(level = "debug", skip_all)]
+async fn process_certificates_helper(
+    response: FetchCertificatesResponse,
+    tx_certificates_loopback: &Sender<CertificateLoopbackMessage>,
+) -> DagResult<()> {
+    trace!("Start sending fetched certificates to processing");
+    if response.certificates.len() > MAX_CERTIFICATES_TO_FETCH {
+        return Err(DagError::TooManyFetchedCertificatesReturned(
+            response.certificates.len(),
+            MAX_CERTIFICATES_TO_FETCH,
+        ));
+    }
```

Comment:
So here we have received a bunch of certificates, that should match the range of the request, and be complete? Should we do any additional checks that the response matches the request before passing them on to the core? For example could a byzantine node send them to us in the wrong order?

Reply:
Yes, I'm relying on the core to handle all byzantine issues e.g. sig mismatch, missing parents here. Duplicating checks here does not seem to get us much additional safety and increases cost at the common case. Since the rate of fetch request / response is limited, the risk of overloading core with junk data should be limited too.